packages/repository/test/acceptance/has-many.relation.acceptance.ts

```diff
@@ -9,52 +9,59 @@ import {
   Entity,
   DefaultCrudRepository,
   juggler,
-  EntityCrudRepository,
-  hasManyRepositoryFactory,
-  HasManyDefinition,
-  RelationType,
+  hasMany,
+  repository,
+  RepositoryMixin,
+  HasManyEntityCrudRepository,
 } from '../..';
 import {expect} from '@loopback/testlab';
+import {inject} from '@loopback/context';
+import {Application} from '@loopback/core';
+import {hasManyRepository} from '../../src/decorators/relation.repository.decorator';
 
 describe('HasMany relation', () => {
   // Given a Customer and Order models - see definitions at the bottom
 
-  beforeEach(givenCrudRepositoriesForCustomerAndOrder);
-
   let existingCustomerId: number;
-  //FIXME: this should be inferred from relational decorators
-  let customerHasManyOrdersRelationMeta: HasManyDefinition;
 
   beforeEach(async () => {
     existingCustomerId = (await givenPersistedCustomerInstance()).id;
-    customerHasManyOrdersRelationMeta = givenHasManyRelationMetadata();
   });
 
   it('can create an instance of the related model', async () => {
-    // A controller method - CustomerOrdersController.create()
-    // customerRepo and orderRepo would be injected via constructor arguments
-    async function create(customerId: number, orderData: Partial<Order>) {
-      // Ideally, we would like to write
-      // customerRepo.orders.create(customerId, orderData);
-      // or customerRepo.orders({id: customerId}).*
-      // The initial "involved" implementation is below
-
-      //FIXME: should be automagically instantiated via DI or other means
-      const customerOrders = hasManyRepositoryFactory(
-        customerId,
-        customerHasManyOrdersRelationMeta,
-        orderRepo,
-      );
-      return await customerOrders.create(orderData);
+    class TestController {
+      constructor(
+        @repository(CustomerRepository) protected cusRepo: CustomerRepository,
+      ) {}
+
+      async createCustomerOrders(
+        customerId: number,
+        orderData: Partial<Order>,
+      ): Promise<Order> {
+        return await this.cusRepo.orders({id: customerId}).create(orderData);
+      }
     }
-
-    const description = 'an order desc';
-    const order = await create(existingCustomerId, {description});
-
+    class TestApp extends RepositoryMixin(Application) {}
+    const app = new TestApp();
+    app.repository(CustomerRepository);
+    app.repository(OrderRepository);
+    app.controller(TestController);
+    app.dataSource(new juggler.DataSource({name: 'db', connector: 'memory'}));
+    const controller = await app.get<TestController>(
+      'controllers.TestController',
+    );
```

Comment:
I am not very happy that we need to create a full application class and use dependency injection to test the new functionality. Isn't there a way how to wire up dependencies explicitly?

For example:

```
const db = new juggler.DataSource({connector: 'memory'});
const controller = new TestController(new CustomerRepository(db));
```

I guess it's ok to have an end-to-end acceptance test that's using an application class and a controller, but it would be good to also have an integration test that does not need dependency injection via `Context`. 

Setting that aside:

I think a truly end-to-end test should invoke controller methods via HTTP/REST.

The current test method is too long and dealing with too many concerns, from setting up an application, registering datasource, retrieving controller/repository classes, etc. Please refactor/extract this code out to helper functions, for example `givenApplicationWithMemoryDb`, `givenCrudRepositoriesForCustomerAndOrder ` (which registers repositories with the app), etc.